avm/src/daemon/ctrl.rs

```diff
@@ -169,25 +188,120 @@ impl ControlPort {
+    });
+    let client = Client::builder().build::<_, Body>(https);
+    //let client = Client::builder().build::<_, Body>(HttpsConnector::from((http_connector, tls)));
+
     ControlPort {
       lrh: LogRendezvousHash::new(vec![]),
-      client: Client::builder().build::<_, Body>(HttpsConnector::from((http_connector, tls))),
+      client,
+      vms: HashMap::new(),
+      self_vm: None,
+      region_vms: HashMap::new(),
+      vms_up: HashMap::new(),
     }
   }
 
-  pub fn update_ips(self: &mut ControlPort, ips: Vec<String>) {
+  pub async fn update_vms(self: &mut ControlPort, self_ip: &str, vms: Vec<VMMetadata>) {
+    let ips = vms
+      .iter()
+      .map(|vm| vm.private_ip_addr.to_string())
+      .collect();
+    let self_vm_vec: Vec<&VMMetadata> = vms
+      .iter()
+      .filter(|vm| vm.private_ip_addr == self_ip)
+      .collect();
+    if self_vm_vec.len() == 0 {
+      error!(
+        ErrorType::NoDnsPrivateIp,
+        "Failed to find self in cluster. Maybe I am being shut down?"
+      )
+      .await;
+      // TODO: Should this error propagate up to the stats loop or no?
+      return;
+    } else if self_vm_vec.len() > 1 {
+      // This hopefully never happens, but if it does, we need to change the daemon initialization
+      error!(
+        ErrorType::DuplicateDnsPrivateIp,
+        "Private IP address collision detected! I don't know who I really am!"
+      )
+      .await;
+      // TODO: Should this error propagate up to the stats loop or no?
+      return;
+    }
+    let self_vm = self_vm_vec[0].clone();
+    let mut region_vms = HashMap::new();
+    vms
+      .iter()
+      .filter(|vm| vm.cloud == self_vm.cloud && vm.region == self_vm.region)
+      .for_each(|vm| {
+        region_vms.insert(vm.private_ip_addr.clone(), vm.clone());
+      });
+    let mut all_vms = HashMap::new();
+    vms.iter().for_each(|vm| {
+      all_vms.insert(vm.private_ip_addr.clone(), vm.clone());
+    });
+    let mut other_region_ips: Vec<String> = region_vms
+      .keys()
+      .filter(|ip| ip.as_str() != self_ip)
+      .filter(|ip| *self.vms_up.get(ip.clone()).unwrap_or(&false))
+      .map(|ip| ip.clone())
+      .collect();
+    let region_ips = Arc::clone(&REGION_VMS);
+    let mut region_ips_mut = region_ips.write().unwrap();
+    region_ips_mut.clear();
+    region_ips_mut.append(&mut other_region_ips);
+    drop(region_ips_mut);
+    drop(region_ips);
+    self.vms = all_vms;
+    self.self_vm = Some(self_vm);
+    self.region_vms = region_vms;
     self.lrh.update(ips);
   }
 
-  pub fn get_leader(self: &ControlPort) -> &str {
-    self.lrh.get_leader_id()
+  pub fn is_leader(self: &ControlPort) -> bool {
+    match &self.self_vm {
+      Some(self_vm) => self.lrh.get_leader_id() == self_vm.private_ip_addr,
+      None => false,
+    }
+  }
+
+  pub fn get_leader(self: &ControlPort) -> Option<&VMMetadata> {
+    match self.vms.get(self.lrh.get_leader_id()) {
+      Some(vm) => Some(&vm),
+      None => None,
+    }
```

Comment:
nit: could be
```suggestion
    self.vms.get(self.lrh.get_leader_id()).as_ref()
```

Reply:
Wouldn't that be `&Option<VMMetadata>` instead of `Option<&VMMetadata>` though?

Reply:
no - `as_ref` inverts the reference so it turns a `&Option<VMMetadata>` into a `Option<&VMMetadata>`